Replace Get function address

Michael-F-Bryan opened the pull request:
## Description

After the discussion in #5 I propose we replace the `ExecutionEngine::get_function_address()` method with a `get_function()` which does the address transmutation for the user.

I believe this is still a fundamentally unsafe operation (there are no guarantees the signature is correct or it's not doing `unsafe` stuff), but at least this version ensures a function pointer doesn't accidentally outlive its parent execution engine.

## How This Has Been Tested

Updated the old tests to use the new API and added an example to the function's documentation which exercises the function.

## Breaking Changes

The `get_function_address()` method has now been replaced with `get_function()`.

TheDan64 commented:
...is this an April fools joke? 😄

TheDan64 commented on src/execution_engine.rs:
```diff
@@ -240,6 +291,22 @@ impl ExecutionEngine {
+/// A wrapper around a function pointer which ensures the symbol being pointed
+/// to doesn't accidentally outlive its execution engine.
+#[derive(Debug, Clone)]
+pub struct Symbol<F> {
```
Is `Symbol` what LLVM calls it? (I do recall a `LLVMGetSymbol` function) I feel like `Function` or even `RuntimeFunction` might be a more descriptive name

TheDan64 commented on src/execution_engine.rs:
```diff
@@ -181,7 +226,13 @@ impl ExecutionEngine {
         }
 
-        Ok(address)
+        assert_eq!(size_of::<F>(), size_of::<usize>(), 
```
This is a neat trick!

TheDan64 replied:
Doesn't transmute already fail to compile if the sizes are different though?

Michael-F-Bryan replied:
Yeahhh....  The `transmute_copy()` lower down makes me feel pretty dirty because it's essentially just a pointer cast. It lets us get around the *"`F` may vary in size"* transmute error, but it's kinda hacky.

TheDan64 replied:
Ok; this is fine. If we find that `Symbol` conflicts with the LLVM idea of `Symbol` we can rename it when we cross that bridge I susppose.

TheDan64 commented on tests/test_builder.rs:
```diff
@@ -207,27 +205,28 @@ fn test_binary_ops() {
-    assert!(xor(false, true));
-    assert!(!xor(true, true));
+    unsafe {
+        type BoolFunc = fn(bool, bool) -> bool;
```
I guess it doesn't matter in this particular instance, but shouldn't the type here be a extern fn?

TheDan64 commented:
This may not actually work, but I wonder if `get_function` doesn't need to be unsafe if we move the logic around. Do you think it's feasible to store the address until it's actually needed, and then do the transmute?

For example:
```rust
Symbol<F> {
    ee: ...,
    address: usize,
}

impl <F> Deref for Symbol {
    type Target = F;

    fn deref(&self) -> &Self::Target {
        let function = unsafe { transmute::<&F>(&self.address) };
        function
    }
}
```

Michael-F-Bryan commented:
> This may not actually work, but I wonder if `get_function` doesn't need to be unsafe if we move the logic around. Do you think it's feasible to store the address until it's actually needed, and then do the transmute?

I really think the `get_function` method *needs* to be `unsafe` to call. We're essentially leaving it up to the user to choose the function signature, so to mark it as safe would probably be unsound.

> This looks reasonable. My remaining question is deref will still return an unsafe function, right?

I believe the `Deref` impl means a `Symbol<F>` will be unsafe if `F` is unsafe (e.g. `unsafe fn()`). There isn't really any way to specify that a function is unsafe to call, so I guess making it unsafe to create a function is about the best we can do :disappointed: 

> I wonder if we could use the nightly `call` from https://doc.rust-lang.org/nightly/std/ops/trait.Fn.html#required-methods ? Seems to be runtime function-ish

I looked at that as well. It'd be pretty much ideal, although I'd like it if there was a way to make the `Fn*` traits `unsafe` to call. I was hesitant to propose implementing the `Fn*` traits though, seeing as it'll confine `inkwell` to be nightly-only, and that's a pain.

Michael-F-Bryan commented:
I think I've found a nice solution. We weren't able to constrain `F` to be `Fn` because that then requires you to name the arguments and then things get kinda annoying.

Instead, I created a "sealed" `UnsafeFunctionPointer` trait and implemented it only for `unsafe extern "C"` functions. 

If you try to retrieve a function pointer which isn't marked `unsafe` or `extern "C"` you now get an error message like this:

```
error[E0277]: the trait bound `fn(): inkwell::execution_engine::UnsafeFunctionPointer` is not satisfied
  --> tests/test_execution_engine.rs:44:34
   |
44 |         assert!(execution_engine.get_function::<fn()>("func").is_ok());
   |                                  ^^^^^^^^^^^^ the trait `inkwell::execution_engine::UnsafeFunctionPointer` is not implemented for `fn()`
   |
   = help: the following implementations were found:
             <unsafe extern "C" fn() -> Output as inkwell::execution_engine::UnsafeFunctionPointer>=
```

It'd be nice to use the [`on_unimplemented`] attribute to give a better error message, but the compiler's suggestions aren't too shabby.

[`on_unimplemented`]: https://github.com/rust-lang/rust/blob/master/src/doc/unstable-book/src/language-features/on-unimplemented.md

TheDan64 commented on src/execution_engine.rs:
```diff
@@ -158,20 +160,73 @@ impl ExecutionEngine {
+    /// unsafe {
+    ///     let test_fn = ee.get_function::<unsafe extern "C" fn() -> f64>("test_fn").unwrap();
+    ///     let return_value = test_fn();
+    ///     assert_eq!(return_value, 64.0);
```
nit: `assert_eq` doesn't need to be in unsafe block. Maybe return it? IE
```rust
let return_value = unsafe {
    let test_fn = ee.get_function::<unsafe extern "C" fn() -> f64>("test_fn").unwrap();
    test_fn()
};
assert_eq!(return_value, 64.0);
```

Unless you think it's less clear that way, feel free to leave it as is.

Michael-F-Bryan replied:
It doesn't really need to be inside the `unsafe` block, but the reason I wrote it that way to show that the `test_fn` and using anything we get from it is innately `unsafe`, and requires a bit of extra attention.

TheDan64 commented on src/execution_engine.rs:
```diff
@@ -253,3 +314,61 @@ impl Drop for ExecutionEngine {
+impl<F> Debug for Symbol<F> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_tuple("Symbol")
+            .field(&"<unnamed>")
```
Maybe we should store the function name from `get_function` for debugging purposes?

Michael-F-Bryan replied:
Hmm... I'm not sure whether that'll be nice to do. We're keeping track of the execution engine by using an `Rc`, and mixing both a reference (with lifetimes) and a reference-counted pointer feels kinda weird.

It also means your function name must outlive the symbol, which isn't really practical or ergonomic. Imagine trying to store a loaded symbol in a struct when the name isn't known until runtime (e.g. the user provided it), it'd be way too easy to get stuck in the rabbit hole that is self-referential types.

TheDan64 replied:
The &str is tied to the input to the function call, not the execution engine though

Michael-F-Bryan commented on src/execution_engine.rs:
```diff
@@ -253,3 +314,61 @@ impl Drop for ExecutionEngine {
+/// to doesn't accidentally outlive its execution engine.
+#[derive(Clone)]
+pub struct Symbol<F> {
+    pub(crate) execution_engine: Rc<LLVMExecutionEngineRef>,
```
I was originally planning to go down the path of embedding an `Rc<ExecutionEngine>` in the `Symbol` tying it back to the original execution engine, but that turned out to be pretty annoying and cumbersome. 

Really, we just want the `Drop` logic that will call the LLVM execution engine destructor when `Rc<LLVMExecutionEngineRef>`'s strong count goes to 1, so I extracted that out into a `ExecEngineInner(Rc<LLVMExecutionEngineRef>)` struct that both `Symbol` and `ExecutionEngine` use.

Michael-F-Bryan commented:
@TheDan64 I've updated the logic around making sure a `Symbol` properly destroys the underlying LLVM execution engine if it's the last person with a reference to it. 

I've also reworked the README example and copied it to the `examples/` directory.

TheDan64 commented on src/execution_engine.rs:
````diff
@@ -89,7 +91,7 @@ impl ExecutionEngine {
     /// assert_eq!(result, 128.);
     /// ```
-    pub fn add_global_mapping(&self, value: &AnyValue, addr: usize) {
+    pub fn add_global_mapping(&mut self, value: &AnyValue, addr: usize) {
````
I wonder if adding mut will just make the API more difficult to use, it doesn't actually add any guarantees since we're pretty much just working with interior mutability anyway...

Michael-F-Bryan replied:
You raise a good point. Initially it felt quite strange to be mutating an execution engine without declaring it as `mut`, but it's probably better to remove the `&mut self` to stay consistent with the rest of the library.

TheDan64 replied:
Yeah, I think we should remove it

TheDan64 commented on src/execution_engine.rs:
```diff
@@ -305,21 +305,47 @@ impl ExecutionEngine {
+    }
+}
+
+impl Clone for ExecutionEngine {
```
Should the EE really get a publicly accessible clone method? `Clone` generally means reallocating a new copy, but here we're just making a new reference to the same EE

TheDan64 commented:
This is coming along nicely. Thank you for all of the effort you've put in! I just have a few more questions and then we should be all set to merge once those are resolved.

TheDan64 commented on src/execution_engine.rs:
```diff
@@ -244,12 +305,96 @@ impl ExecutionEngine {
+}
 
-        if Rc::strong_count(&self.execution_engine) == 1 {
+impl Clone for ExecutionEngine {
```
Should the EE really get a publicly accessible clone method? Clone generally means reallocating a new copy, but here we're just making a new reference to the same EE. Feels a bit counter intuitive to me

Michael-F-Bryan replied:
We end up using the equivalent of `Clone` in a couple places around the code base to create a new reference to the same underlying LLV execution engine so all I did was extract this into an impl for the `Clone` trait. 

I think people will understand this better if we mention an `ExecutionEngine` is actually just a reference-counted pointer to an underlying LLVM execution engine. But I don't mind deleting (or hiding with `#[doc(hidden)]`) this `Clone` impl, it's up to you really.

TheDan64 replied:
Feel free to add documentation for it; preferably on the clone implementation itself (assuming rustdoc will make this visible, otherwise on the EE itself I guess)

Michael-F-Bryan replied:
Rustdoc lets you override the docs for trait implementations, but I'll probably mention that an `ExecutionEngine` is a reference-counted pointer anyway.

Michael-F-Bryan commented:
@TheDan64 I think I've addressed all our previous concerns and just rebased onto `master`, would you be able to do another review and let me know if there's anything else we need to do for this PR?

TheDan64 commented:
Sure thing, I should have time to do a final review this weekend
